### DISCUSSION TITLE
Can we programatically extract % of each class from .hdf5 output file?

### DISCUSSION THREAD

**Discussion by lomshabhishek** [AICAN-Research/FAST-Pathology#92] [General]
I ran BACH classifier and was wondering if we can programmatically extract % of each class from the .hdf5 generated by the fastpathology desktop app

**Comment by andreped** [comment]
Hello, @lomshabhishek! :] 

**NOTE:** Next time I recommend making a Q/A instead of a General post, as we can assign solution to the original question.

Regarding your question, you can do that from python. In FP, you should have created a project, added an image to the project, and ran the BACH model, right? In that case, the generated HDF5 file should lie in the project.

Given the path of the HDF5 file, you can then read it by:
```
import h5py
import numpy as np

with h5py.File("/path/to/file.hdf5", "r") as f:
    data = np.asarray(f["tensor"])

print(data.shape)

argmaxed = np.array(data, axis=-1)
print(argmaxed.shape)
print(np.unique(argmaxed, return_counts=True))
```

The tensor should be an image with four channels. I then argmaxed it across the channel dimension which should assign one class to each pixel. I then return the class frequency. If you are cool, you could also just calculate the percentage directly using:
```
print(np.histogram(argmaxed, density=True))
```

Just note that most of that image will be empty (zero value), as most of the image will be glass. In that case, if one of your classes is assigned to 0, which is a non-glass class, this class frequency will be overestimated, as both the class of interest and background will be included in this count.

So if you want to only calculate the percentage based on tissue, you will first need to get the tissue image, and then remove it from the BACH image by doing something like:
```
argmaxed = argmaxed.flatten()
tissue = tissue.flatten()
argmaxed = argmaxed[tissue == 1]
print(np.histogram(argmaxed, density=True)
```

The tissue image you can save in the project and load it. You might need to resize it to match the heatmap (HDF5) output. When resizing remember to use nearest neighbour to avoid any funny business.

Maybe this is enough for you to get started. Let me know if you have any other issues :]

**Reply by lomshabhishek** [reply]
Tissue output is a tiff file and if I compare their shape after flatten then i get -> (13348,) (3417088,) for heatmap and tisse image respectively 

I am using this as you mentioned, it seems to be working. Thanks.

```
tissue_image = Image.open(tissue_image_path)
tissue_image_resized = tissue_image.resize((heatmap_data.shape[1], heatmap_data.shape[0]), Image.NEAREST)
```

**Reply by lomshabhishek** [reply]
```
NOTE: Next time I recommend making a Q/A instead of a General post, as we can assign solution to the original question.
```

Sure noted, thanks again for this solution, I can raise this in Q/A as well if its fine with you.

**Reply by andreped** [reply]
> Sure noted, thanks again for this solution, I can raise this in Q/A as well if its fine with you.

No need. Just do it next time if you have a question, as it should be easier for others to find the `solution` to the question :]

> Tissue output is a tiff file and if I compare their shape after flatten then i get -> (13348,) (3417088,) for heatmap and tisse image respectively
> 
> I am using this as you mentioned, it seems to be working. Thanks.

It is true that you need to resize the tissue image to the same **_scale_** as the heatmap, but with the current solution you will run into an **edge issue**.

Note that during patch generation, some extracted patches will be from the edge of the WSI (right and bottom) where parts of the patch may lie outside the WSI. In this case, we (zero)pad the patch to match the desired input of the network. However, when stitching the output of the network to then form the heatmap, we end up with a heatmap that cover a small part that goes outside the WSI. In this case, by resizing the tissue segmentation to exactly match the heatmap is size, will result in a small distortion, where the one-to-one pixel relation won't be exact. If you open the project in FastPathology with the generated heatmap on a WSI and zoom in on the right-hand size of the WSI, it might be easier to understand what the problem is.

May be that this error is OK for your application. If not, sadly the solution is a bit tricky. I do not have time to sketch a solution for you, but may take a look in January when I am back.

**Reply by lomshabhishek** [reply]
It works for my experiment, Thanks.

**Comment by andreped** [comment]
@smistad Isn't this a very natural feature people would be interested in seeing when running a WSI-classifier using a patch-wise classifier, or just a patch-wise classifier for segmentation in general? Adding support for this for pw-classifiers (heatmaps) should be a lot easier than for semantic segmentation models.

If we are able to accumulate as the pw-classifier runs to build up a histogram, when it is finished, it could be something that is returned. In FP, we could save this histogram on disk in CSV format, or even easier, as part of the already existing HDF5 similar to `scaling`. In FP, we could also display the bincount somehow in the View widget.

I believe we played around with adding support for [qwt](https://qwt.sourceforge.io/) or something similar. It might even been added support for building FAST with said library, but I am not sure this was ever tested. I do not really recall atm which it was. Do you, @smistad?

IMO, the ideal library would be `QtCharts`, but it is GPL-licensed, so it is a no-go for FAST and FP.

Shall I make a feature request under issues in FP?

**Reply by smistad** [reply]
Yes, that could be. But we would need more resources to implement it.
Another solution can be to provide some pyFAST examples/tutorials on how to work with the stored output from FP.

PS: Please don't close the discussion as it makes it harder for others to find it. I am not sure if we can disable the possibility of closing discussions on github...

**Reply by andreped** [reply]
> Another solution can be to provide some pyFAST examples/tutorials on how to work with the stored output from FP.

I have actually done something very similar, but I remember it being a little bit more tricky to do than first anticipated. Due to the heatmap covering some parts outside the WSI. I may have time this weekend to make a simple notebook on how to do this.
